iothub/device/src/Transport/Amqp/AmqpConnectionPool.cs

```diff
@@ -99,8 +84,16 @@ public void RemoveAmqpUnit(AmqpUnit amqpUnit)
                 {
                     AmqpConnectionHolder[] amqpConnectionHolders = ResolveConnectionGroup(deviceIdentity);
                     amqpConnectionHolder = ResolveConnectionByHashing(amqpConnectionHolders, deviceIdentity);
+
+                    amqpConnectionHolder.RemoveAmqpUnit(amqpUnit);
+
+                    // If the connection holder does not have any more units, the entry needs to be nullified.
+                    if (amqpConnectionHolder.IsEmpty())
+                    {
+                        int index = GetDeviceIdentityIndex(deviceIdentity, amqpConnectionHolders.Length);
+                        amqpConnectionHolders[index] = null;
```

Comment:
why do we need to do this?

Reply:
So when the last amqp unit gets removed, the connection pool has a blank slate.